### DISCUSSION TITLE
How hard would Linux support be?

### DISCUSSION THREAD

**Discussion by rizerco** [groue/GRDB.swift#1359] [General]
I’ve recently started using GRDB, but now I’m starting to consider how to take some of the core parts of my app cross platform. What are the main blockers for Linux support? Would it be the same for Windows? Is it a case of a few minor changes and `#if canImport`s or a major rewrite to certain parts?
Thanks!
Luke

**Comment by groue** [comment]
Hello @rizerco,

Thanks for your interest in GRDB!

I do not know how far we are from Linux support.

Actually there's three levels:

1. Have your fork of GRDB build and run on Linux.
2. Have your changes merged upstream.
3. Have GRDB claim Linux support.

The first level is up to you.

The second level will quickly rot as time passes. I will merge changes that break Linux support eventually.

The third level adds a few requirements, because **someone** will have to keep it working on Linux, **over the years**, and this someone will be **me**. I do not run any Linux box, and this means I'll need a fair amount of help:

1. The CI scripts must be updated to add Linux tests.
2. There must be an easy way for me to run and **debug** GRDB in some virtualized container on my Mac.
3. This must keep on running smoothly as macOS and Linux distros ship new releases.

This is risky, because there's not much demand for Linux support, and I don't personally use it. I will not hesitate stopping the Linux experiment if things turn too difficult/painful/time consuming.

There's a safer option, where **you** are the long-term maintainer of the GRDB fork that adds Linux support. As time passes, we'll have more discussions, get to know each other, and some of your changes will be merged in the main repo, in a warm and fruitful collaboration.

In all cases, everything starts at level 1, described above :-) You'll find some starting points in [CONTRIBUTING](https://github.com/groue/GRDB.swift/blob/master/CONTRIBUTING.md).

**Reply by rizerco** [reply]
Thanks! I got the package itself to build on Linux without a huge amount of effort, although it looks like it will take a bit longer to get the tests to build. I think I need to consider how much time I want to invest in making this happen, but at least it’s a promising start.

**Reply by groue** [reply]
🙂 Yes, indeed! The makefile contains tests for SPM: maybe you can get some inspiration there?
